File: trees/c++/04_1.cpp
/*
 * create a mirror tree of a binary tree
 *
 * */
#include<iostream>

struct node
{
    int data;
    node *left, *right;
};

node* get_new_node(int data)
{
    node* new_node = new node();
    new_node->data = data;
    new_node->left = new_node->right = NULL;
    return new_node;
}

node* insert(node* root, int data)
{
    // case 1 when root is null
    if (!root)
        root = get_new_node(data);
    // case 2 when data is less than root
    else if (data < root->data)
        root->left = insert(root->left, data);
    // case 3 when data is greater than root
    else
        root->right = insert(root->right, data);
    return root;
}

void swap(node** a, node** b)
{
    node *temp = *a;
    *a = *b;
    *b = temp;
}

node* get_mirror_tree(node* root)
{
    if (!root)
        return root;
    else
    {
        get_mirror_tree(root->left);
        get_mirror_tree(root->right);
        swap(&root->left, &root->right);
    }
}

void inorder(node* root)
{
    if (root)
    {
        inorder(root->left);
        std::cout<<root->data<<" ";
        inorder(root->right);
    }
}

int main()
{
    node *root1 = NULL, *root2 = NULL;

    root1 = insert(root1, 50);
    root1 = insert(root1, 40);
    root1 = insert(root1, 60);
    root1 = insert(root1 ,30);
    root1 = insert(root1, 35);
    root1 = insert(root1, 70);
    root1 = insert(root1, 20);
    root1 = insert(root1, 10);

    std::cout<<std::endl<<std::endl;

    inorder(root1);

    get_mirror_tree(root1);


    std::cout<<std::endl;
    inorder(root1);
    return 0;
}

//T.C = O(n)
//S.C = O(n)

File: trees/python/04_1.py
import bst

'''
if you observe, you will find out that we just need to swap left and right
child of each node in the tree and thus to implement this we can use recursion
easily.

So for every node we just need to swap the left sub tree and right sub tree.
Thus, we can the function recursively with left sub tree and right sub tree
and as we reach to the leaf node(or last but leaf node), we can swap their
child (i.e. two null pointer), then that recursion stack will get destroyed
then their parent node's children will get swapped
'''


'''
this method isn't returning the root, because the root has been passed as
reference (actually by value, where the value passed is reference itself)
Also, note that I am not rebinding the object here
'''


def indent(n):
    for i in range(n):
        print '   ',


def getMirrorTree(root, inden):
    indent(inden)
    print 'getMirrorTree('+str(root)+')'
    if not root:
        return
    if root.left is not None:
        getMirrorTree(root.left, inden+1)
    if root.right is not None:
        getMirrorTree(root.right, inden+1)
    root.left, root.right = root.right, root.left


def getMirror(root):
    getMirrorTree(root, 0)


if __name__ == '__main__':
    root = None
    BSTree = bst.BST()
    inp = input('enter the elements of tree separated by ,\n')
    for i in inp:
        root = BSTree.insert(root, i)

    print 'inorder traversal of original tree\n'
    BSTree.inorder_rec(root)
    print

    getMirror(root)

    print
    print 'inorder traversal of mirror tree\n'
    BSTree.inorder_rec(root)

# time complexity = O(n)
# UPDATE: I have optimized the previous solution, previousily I was calling
# the recursive function even when the root.left or root.right was None, this
# was leading to the swapping of None and None.
